Feature/atom checkbox

devictoribero opened the pull request:
# Checkbox component

## Assumptions
- Input isolated, without labels.
- Since we had no sizes in the PIM, I preferred not to add (since I don't know if we will use them. We can always iterate)
- The component must be accessible

## What does this PR includes?
- **Source code** of the `Checkbox` component
- **Tests** that validate the correct functionality of the component
- **A story** to [see how this component is rendered go to this link](https://606df1b4143a4e00218b839d-iandrjmnzi.chromatic.com/)


## Designs
**primary**
<img width="208" alt="Screenshot 2021-05-11 at 10 33 56" src="https://user-images.githubusercontent.com/16169890/117852341-75466500-b244-11eb-8639-eb493d1c90de.png">

**secondary** ( ⚠️ we need to improve the accessibility of our color palette)
<img width="213" alt="Screenshot 2021-05-11 at 10 33 52" src="https://user-images.githubusercontent.com/16169890/117852507-9eff8c00-b244-11eb-8856-6142dfe154a3.png">


**small**
<img width="165" alt="Screenshot 2021-05-11 at 10 34 01" src="https://user-images.githubusercontent.com/16169890/117852421-898a6200-b244-11eb-9304-ebe963cc0b71.png">

**medium**
<img width="204" alt="Screenshot 2021-05-11 at 10 34 04" src="https://user-images.githubusercontent.com/16169890/117852463-9444f700-b244-11eb-817b-7033984d9e1b.png">

**large**
<img width="238" alt="Screenshot 2021-05-11 at 10 34 10" src="https://user-images.githubusercontent.com/16169890/117852545-a6bf3080-b244-11eb-9da0-4b1b58e6391f.png">

devictoribero commented:
Should we use the current `frontend` team we have in the organization? or should we create a new one instead? we could add that as default so everyone gets noticed. The more automated the better.

What do you think @hakonkrogh @danisal @deeksha-crys ?

hakonkrogh commented:
Is there a storybook preview for this?

devictoribero commented:
As @deeksha-crys pointed out on slack, we might want to add the `color` and the `size` props even though we don't use them now and don't have a specification either.

The reason is that these props will be common for all/most of the components. So, I'm going to add these.

devictoribero commented:
Added the `size` and `color` props as @deeksha-crys  mentioned

devictoribero commented on src/components/Checkbox.tsx:
```diff
@@ -0,0 +1,170 @@
+  id,
+  name,
+  onChange,
+  size = checkboxSizes.medium,
```
Made `checkboxSizes.medium` the default because it's the one we are using on the PIM at this moment, but to me, it looks more natural to make the default size 14px. What do you think?

devictoribero commented:
The demo is available at https://606df1b4143a4e00218b839d-iandrjmnzi.chromatic.com/

danisal commented:
When the size is undefined, I guess the checkboxes in this size... 

![image](https://user-images.githubusercontent.com/9869755/118149659-a754f400-b409-11eb-90fa-bfa9c4a4d4bf.png)

I guess the `undefined` should have a default, like small or medium

danisal commented:
Also, when I set the disabled to true, it still works, I don't know if that is just a storybook thing or not

![image](https://user-images.githubusercontent.com/9869755/118149988-ff8bf600-b409-11eb-8218-19cfdb79bc1b.png)

danisal commented:
And when the color is set to  `undefined` this happens

![image](https://user-images.githubusercontent.com/9869755/118150107-16324d00-b40a-11eb-9217-7c4305d947fd.png)

It might be just a storybook thing. I don't know

deeksha-crys commented on package.json:
```diff
@@ -26,7 +27,7 @@
   "husky": {
     "hooks": {
-      "pre-commit": "tsdx lint"
+      "pre-commit": "tsdx lint && tsdx test"
```
@devictoribero nope this is not working yet.

deeksha-crys commented:
@devictoribero changing the `color` prop on Chromatic story is throwing `undefined`. If you handle the case when the `size` is large and no color is provided (show default color) then it would work.

deeksha-crys commented:
So `undefined` on the Chromatic story means when the specific prop is not provided. It mimics the default case for the component. If that is handled in the code, then it would be reflected here in Chromatic.

deeksha-crys commented:
Quick question, I see that Chromatic is missing the Typography component. Did any refactoring remove the existing component in this branch?

deeksha-crys commented:
When I am applying multiple props together in the Chromatic, they don't seem to be working. Not sure if I am using the correct Chromatic link or the state is a bit shaky.

So what I have learned so far is that on Chromatic canvas, we can use the component the way we do in the code. For example <Checkbox size="medium" disabled={false}/>. We can visually see what happens when someone will use this component in the code. So maybe when you add a new prop, you can quickly test if that is working with other props too. It's like a state machine/Automata. 

However additionally what I did was, publish this library on NPM through CLI (and delete it later) 
or even from my local directory `yarn install ../mylib` and use it inside a consumer app (a CRA which you can spin up locally). I was able to fix quite a few bugs during this approach.

devictoribero commented:
@deeksha-crys I've updated the component as we talked. I added the `color` and `size` prop. Also, I have tested it in a separate repository and it works fine.

In addition, I have tried the issue @danisal was pointing out about the `disabled: true` still working, and it's an issue with storybook because the `onChange` callback is never executed when is disabled.

I tested it on the `pim` but looks like something is broken, I will review that tomorrow.

hakonkrogh commented:
:tada: This PR is included in version 1.4.0 :tada:

The release is available on:
- [npm package (@beta dist-tag)](https://www.npmjs.com/package/@crystallize/design-system/v/1.4.0)
- [GitHub release](https://github.com/CrystallizeAPI/design-system/releases/tag/v1.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
